### PR TITLE
ARROW-15980: DataFusion-Contrib Announcement Blog Post

### PR DESCRIPTION
**Pull request by matthewmturner**

### CODE REVIEW

File: _posts/2022-03-15-datafusion-contrib.md [hunk 1]
```diff
@@ -0,0 +1,90 @@
+---
+layout: post
+title: Intorducing Apache Arrow DataFusion Contrib
```
Comment (andygrove) [inline _posts/2022-03-15-datafusion-contrib.md:3]:
```suggestion
title: Introducing Apache Arrow DataFusion Contrib
```

**Comment by houqp** [comment]
this looks like a very useful catalog to be linked from datafusion's own home page as well :+1:

**Comment by matthewmturner** [comment]
@houqp great idea, i can handle that as part of my work on refreshing docs.

File: _posts/2022-03-16-datafusion-contrib.md [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+As context, the optimizer framework within DataFusion is currently under review with the objective of implementing a more strategic long term solution that is more efficient and simpler to develop.
+
+Some of the benefits of using `egg` within DataFusion are:
+
+- An optimized algorithm that is hard to match with manually written optimization passes
+- Easier and less verbose to add simple rules
+- Plugin framework to add slightly more complex optimizations
+- Does not depend on rule order and combined with being able to apply multiple rules in one pass and until convergence, can optimize further than currently is possible (or what would be possible using a optimization strategy like Apache Spark)
+- Cost based optimizations are easy to add
```
Comment (Dandandan) [inline _posts/2022-03-16-datafusion-contrib.md:134]:
```suggestion
- Implements optimized algorithms that are hard to match in terms of with manually written optimization passes
- Makes it easy and less verbose to add optimization rules
- Plugin framework to add more complex optimizations
- Egg does not depend on rule order and can lead to a higher level of optimization by being able to apply multiple rules at the same time until it converges
- Allows for cost-based optimizatios
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+{% endcomment %}
+-->
+
+# Introduction
+
+[DataFusion](https://arrow.apache.org/datafusion/) is an extensible query execution framework, written in Rust, that uses Apache Arrow as its in-memory format.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:29]:
```suggestion
Apache Arrow [DataFusion](https://arrow.apache.org/datafusion/) is an extensible query execution framework, written in Rust, that uses [Apache Arrow](https://arrow.apache.org) as its in-memory format.
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+When you want to extend your Rust project with [SQL support](https://arrow.apache.org/datafusion/user-guide/sql/sql_status.html), a DataFrame API, or the ability to read and process Parquet, JSON, Avro or CSV data, DataFusion is definitely worth checking out.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:31]:
```suggestion
When you want to extend your Rust project with [SQL support](https://arrow.apache.org/datafusion/user-guide/sql/sql_status.html), a DataFrame API, or the ability to read and process Parquet, JSON, Avro or CSV data, DataFusion is definitely worth checking out. DataFusion's pluggable design makes creating extensions at various points particular easy to build.
```

Comment (alamb) [reply _posts/2022-03-16-datafusion-contrib.md:31]: I thought maybe setting up the "DataFusion is extensible" would motivate the project some more as well as highlight another benefit of DataFusion

File: _posts/2022-03-16-datafusion-contrib.md [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+
+DataFusion's  SQL, `DataFrame`, and manual `PlanBuilder` API let users access a sophisticated query optimizer and execution engine capable of fast, resource efficient, and parallel execution that takes optimal advantage of todays multicore hardware. Being written in Rust means DataFusion can offer *both* the safety of dynamic languages as well as the resource efficiency of a compiled language.
+
+The Apache Arrow DataFusion team is pleased to announce the creation of the [DataFusion-Contrib](https://github.com/datafusion-contrib) GitHub organization.  This organization was created to provide an unofficial testing ground for new DataFusion features and as home for extensions to the core DataFusion project.  There are currently five active repositories within this organization which are summarized below.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:35]:
```suggestion
The DataFusion team is pleased to announce the creation of the [DataFusion-Contrib](https://github.com/datafusion-contrib) GitHub organization to support and accelerate other projects.  While the core DataFusion library remains under Apache governance, the contrib organization provides a more flexible testing ground for new DataFusion features and a home for DataFusion extensions.

There are currently several active repositories within this organization which are summarized below.
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+
+## DataFusion-Python
+
+This [project](https://github.com/datafusion-contrib/datafusion-python) provides Python bindings to the core Rust implementation of DataFusion.  It already provides many core methods for working with DataFusion such as:
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:39]:
```suggestion
This [project](https://github.com/datafusion-contrib/datafusion-python) provides Python bindings to the core Rust implementation of DataFusion, which allows users to:
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+
+- Work with a SQL or DataFrame API for creating query plans, executing them in a multi-threaded environment, and returning results in Python
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:41]:
```suggestion
- Work with familiar SQL or DataFrame APIs to run queries in a safe, multi-threaded environment, returning results in Python
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+- Creation of User Defined Functions and User Defined Aggregate Functions for complex operations
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:42]:
```suggestion
- Create User Defined Functions and User Defined Aggregate Functions for complex operations
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+- Provides zero copy between Python and underlying Rust based execution engine
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:43]:
```suggestion
- Pay no overhead to copy between Python and underlying Rust execution engine (by way of Apache Arrow arrays)
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+
+### Upcoming enhancements
+
+Going forward there will be a focus on exposing more features from the underlying Rust implementation of DataFusion and improving documentation.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:47]:
```suggestion
The team is focusing on exposing more features from the underlying Rust implementation of DataFusion and improving documentation.
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+Or
+
+```bash
+python -m pip install datafusion
+```
+
+## DataFusion-ObjectStore-S3
+
+This [crate](https://github.com/datafusion-contrib/datafusion-objectstore-s3) provides an `ObjectStore` implementation for querying data stored in S3 or S3 compatible storage.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:65]:
```suggestion
This [crate](https://github.com/datafusion-contrib/datafusion-objectstore-s3) provides an `ObjectStore` implementation for querying data stored in S3 or S3 compatible storage. This makes it almost as easy to query data that lives on S3 as lives in local files
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+
+Add the below to your `Cargo.toml` in your Rust Project with DataFusion.
+
+```toml
+datafusion-objectstore-s3 = "0.1.0"
+```
+
+## DataFusion-Substrait
+
+Substrait provides a cross-language serialization format for relational algebra, based on protocol buffers.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:84]:
```suggestion
[Substrait](https://substrait.io/) is an emerging standard that provides a cross-language serialization format for relational algebra (e.g. expressions and query plans).
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+This [crate](https://github.com/datafusion-contrib/datafusion-substrait) provides a Substrait producer and consumer for DataFusion.  A producer converts a DataFusion logical plan into a Substrait protobuf and a consumer does the reverse.
+
+Examples of how to use this crate can be found [here](https://github.com/datafusion-contrib/datafusion-substrait/blob/main/src/lib.rs).
+
+### Potential Use Cases
+
+- Replace current DataFusion protobuf definition used in Ballista for passing query plan fragments to executors.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:92]:
```suggestion
- Replace custom DataFusion protobuf serialization.
- Ecosystem interoperability: DataFusion can run plans produced by other query engines, and visa versa
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+
+## DataFusion-HDFS
+
+This [crate](https://github.com/datafusion-contrib/datafusion-objectstore-hdfs) introduces `HadoopFileSystem` as a remote `ObjectStore` which provides the ability to query HDFS files.  For HDFS access the [fs-hdfs](https://github.com/yahoNanJing/fs-hdfs) library is used.
+
+## DataFusion-Tokomak
+
+This [crate](https://github.com/datafusion-contrib/datafusion-tokomak) provides an e-graph based DataFusion optimization framework based on the Rust [egg](https://egraphs-good.github.io) library.  An e-graph is a data structure that powers the equality saturation optimization technique.
+
+As context, the optimizer framework within DataFusion is currently under review with the objective of implementing a more strategic long term solution that is more efficient and simpler to develop.
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:126]:
```suggestion
As context, the optimizer framework within DataFusion is currently [under review](https://github.com/apache/arrow-datafusion/issues/1972) with the objective of implementing a more strategic long term solution that is more efficient and simpler to develop.
```

File: _posts/2022-03-16-datafusion-contrib.md [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+- A new help tab that provides information on functions
+- Query custom `TableProvider`s such as [DeltaTable](https://github.com/delta-io/delta-rs) or [BigTable](https://github.com/datafusion-contrib/datafusion-bigtable)
+
+## DataFusion-Java
+
+This [project](https://github.com/datafusion-contrib/datafusion-java) created an initial set of Java bindings to DataFusion.  The project is currently in maintenance mode and is looking for maintainers to drive future development.
+
+# How to Get Involved
+
+If you are interested in contributing to DataFusion, and learning about state of
```
Comment (alamb) [inline _posts/2022-03-16-datafusion-contrib.md:170]: I feel like we should plug / suggest how to add new community contributions here too 🤔

**Review by alamb — APPROVED** [review]
Very cool @matthewmturner  Thank you to everyone who has contributed to this

**Review by alamb — COMMENTED** [review]
Any other thoughts on this blog post before we publish it?

**Comment by matthewmturner** [comment]
I just added small intro text, updated date to tomorrow (3/21 - assuming we publish then), and added datafusion-streams section.  good to go from my end.

**Review by alamb — APPROVED** [review]

**Comment by alamb** [comment]
🚀

**Pull request merged**